client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -379,6 +362,18 @@ public void appendSetup(AppendSetup appendSetup) {
             }
         }
 
+        private void retryOnWriteFail(WireCommand wireCommand) {
+            if (state.encounteredSeal()) {
```

Comment:
One clarification: we only execute the retry loop in the case the state implies that the segment is sealed, but the situation in `noSuchSegment` isn't the one of a segment sealed, instead, the segment is gone due to truncation. Are we checking for the right state here? Are we overloading segment sealed to also mean deleted segments due to truncation?

Reply:
SegmentOutputStream changes `SegmentOutputStreamImpl.State` is to `Seal`  and invokes the callback from EventStreamWriter which resends all the inflight events to the correct segments and does an flush on every segment.
The behavior on receiving `noSuchSegment` and `segmentIsSealed` is same, all the inflight events must be re-sent to the correct segments.

Reply:
I understand that the behavior is the same as in both cases the writer needs to go after the successors, but the naming is odd. Refactoring should be done in a separate PR, but I suggest we change the naming to reflect correctly that this not only about sealing, it is more like state == `needSuccessors`.

Reply:
done, updated it to `state.needSuccessors`.